### PR TITLE
diagnostic_tool: Add the sdn check

### PR DESCRIPTION
**Pull request by Miciah**

### CODE REVIEW

**Review by pravisankar — CHANGES_REQUESTED** [review]

File: admin_guide/diagnostics_tool.adoc [hunk 1]
```diff
@@ -298,6 +298,11 @@ A user-defined timeout may be set by passing the
 if a newly-created log entry is not able to be queried via Elasticsearch after
 45 seconds.
 
+|`sdn`
+|This check performs some cluster-level diagnostics of the OpenShift SDN.  In particular, it verifies that master hosts can connect to kubelets and that nodes can route packets to each other; validates node addresses and HostSubnet objects; and gathers networking-related objects from the OpenShift API and logs, OVS flows, iptables rules, and other network configuration from master and node hosts that may help to diagnose problems with the SDN.
+
+This Ansible-based check complements the `oc adm diagnostics NetworkCheck` diagnostic: the former is oriented towards diagnosing problems at the infrastructure or node level whereas the latter is oriented towards diagnosing problems at the application or pod level.
```
[the remark this answers is not in the document]
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Thanks! Fixed.

File: admin_guide/diagnostics_tool.adoc [hunk 2]
```diff
@@ -298,6 +298,11 @@ A user-defined timeout may be set by passing the
...
+
```
[the remark this answers is not in the document]
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Thanks! Updated.  Does the new text look all right?

**Review by pravisankar — APPROVED** [review]
/lgtm

**Comment by pravisankar** [comment]
@openshift/team-documentation PTAL

File: admin_guide/diagnostics_tool.adoc [hunk 3]
```diff
@@ -298,6 +298,11 @@ A user-defined timeout may be set by passing the
 if a newly-created log entry is not able to be queried via Elasticsearch after
 45 seconds.
 
+|`sdn`
+|This check performs some cluster-level diagnostics of the OpenShift SDN.  In particular, it verifies that master hosts can connect to kubelets and that nodes can route packets to each other; validates node addresses and HostSubnet objects; and gathers networking-related objects from the OpenShift API and logs, OVS flows, iptables rules, and other network configuration from master and node hosts that may help to diagnose problems with the SDN.
+
+This Ansible-based check complements the `oc adm diagnostics networkcheck` diagnostic: the former is oriented towards diagnosing problems at the infrastructure or node level whereas the latter is oriented towards diagnosing problems at the application or pod level.  Use `oc adm diagnostics networkcheck` when the master is able to schedule pods on nodes but pods have connectivity problems.  Use the Ansible check if the cluster is in a state such that diagnostic pods cannot be scheduled or do not provide sufficient information to troubleshoot a problem.
```
Comment (kalexand-rh) [inline admin_guide/diagnostics_tool.adoc]: I might replace this paragraph with "This check can help you diagnose pod or infrastructure problems, particularly if you cannot schedule diagnostic pods or they do not provide enough information to troubleshoot the issue."

At line 118, I'd say, "Create diagnostic pods on multiple nodes to diagnose common network issues from an application or pod standpoint. Run this diagnostic when when the master can schedule pods on nodes, but the pods have connection issues. This check confirms that pods can connect to services, other pods, and the external network."

**Review by kalexand-rh — COMMENTED** [review]
I have some suggestions and requests. Please wrap your lines to 80 characters.

File: admin_guide/diagnostics_tool.adoc [hunk 4]
```diff
@@ -298,6 +298,11 @@ A user-defined timeout may be set by passing the
 if a newly-created log entry is not able to be queried via Elasticsearch after
 45 seconds.
 
+|`sdn`
+|This check performs some cluster-level diagnostics of the OpenShift SDN.  In particular, it verifies that master hosts can connect to kubelets and that nodes can route packets to each other; validates node addresses and HostSubnet objects; and gathers networking-related objects from the OpenShift API and logs, OVS flows, iptables rules, and other network configuration from master and node hosts that may help to diagnose problems with the SDN.
```
[the remark this answers is not in the document]
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Done, done, done, done, done, and done (by turning it into a ul).  Thanks!

File: admin_guide/diagnostics_tool.adoc (hunk 3 above)
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Changes made.  Thanks!

File: admin_guide/diagnostics_tool.adoc [hunk 5]
```diff
@@ -298,6 +302,24 @@ A user-defined timeout may be set by passing the
...
+|`sdn`
+a|This check performs the following cluster-level diagnostics of the
+{product-title} SDN:
+
+* Verify that master hosts can connect to kubelets.
+* Verify that nodes can route packets to each other.
+* Validate node addresses.
+* Validate HostSubnet objects.
+
+If you specify the `*openshift_checks_output_dir*` variable, the check will also
```
Comment (kalexand-rh) [inline admin_guide/diagnostics_tool.adoc]: s/will also save/also saves
Everything else LGTM. :)

Which version(s) is this change valid for @Miciah? (If you have a BZ or Trello with the info, you could link to it.)

@openshift/team-documentation, PTAL

Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Fixed, thanks!

This is for OCP 3.11 onwards.

https://github.com/openshift/openshift-ansible/pull/7759
https://trello.com/c/HtaFZbiR
No Bugzilla report.

Comment (kalexand-rh) [reply admin_guide/diagnostics_tool.adoc]: Thank you for being so responsive! I'll hold onto this one until we open the 3.11 docs, then. :)

File: admin_guide/diagnostics_tool.adoc [hunk 6]
```diff
@@ -115,7 +115,11 @@ Available diagnostics include:
 |Check that the integrated Heapster metrics can be reached via the cluster API proxy.
 
 |`NetworkCheck`
-|Create diagnostic pods on multiple nodes to diagnose common network issues from an application standpoint. For example, this checks that pods can connect to services, other pods, and the external network.
+|Create diagnostic pods on multiple nodes to diagnose common network issues from
+an application or pod standpoint. Run this diagnostic when when the master can
```
Comment (mburke5678) [inline admin_guide/diagnostics_tool.adoc]: s/when when/when

File: admin_guide/diagnostics_tool.adoc [hunk 7]
```diff
@@ -298,6 +302,24 @@ A user-defined timeout may be set by passing the
...
+|`sdn`
+a|This check performs the following cluster-level diagnostics of the
+{product-title} SDN:
+
+* Verify that master hosts can connect to kubelets.
+* Verify that nodes can route packets to each other.
+* Validate node addresses.
+* Validate HostSubnet objects.
+
+If you specify the `*openshift_checks_output_dir*` variable, the check also
```
Comment (mburke5678) [inline admin_guide/diagnostics_tool.adoc]: Where to specify `openshift_checks_output_dir`?

File: admin_guide/diagnostics_tool.adoc [hunk 8]
```diff
@@ -298,6 +302,24 @@ A user-defined timeout may be set by passing the
...
+saves networking-related objects from the {product-title} API as well as logs,
+OVS flows, iptables rules, and other network configuration information under the
+specified directory.
+
+This check can help you diagnose pod or infrastructure problems when you cannot
+schedule diagnostic pods or they do not provide enough information to
+troubleshoot the issue.
+
```
Comment (mburke5678) [inline admin_guide/diagnostics_tool.adoc]: s/they/the diagnostic pods ? 
Not 100% clear and seems a bit awkward to me.

**Comment by mburke5678** [comment]
@kalexand-rh A couple of nits and a question

File: admin_guide/diagnostics_tool.adoc (hunk 6 above)
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Fixed.  Thanks!

File: admin_guide/diagnostics_tool.adoc (hunk 7 above)
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: I added a sentence referring to the example usage below.  Is that sufficient, or would it be preferable to add an explicit example of setting `openshift_checks_output_dir` in the usage examples?

File: admin_guide/diagnostics_tool.adoc (hunk 8 above)
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: I reworded it a little; is it clearer and less awkward now?

File: admin_guide/diagnostics_tool.adoc (hunk 7 above)
Comment (mburke5678) [reply admin_guide/diagnostics_tool.adoc]: @Miciah It is not clear to me where the user sets the variable.

File: admin_guide/diagnostics_tool.adoc (hunk 8 above)
Comment (mburke5678) [reply admin_guide/diagnostics_tool.adoc]: @Miciah Maybe: _or the diagnostic pods do not provide enough information_

File: admin_guide/diagnostics_tool.adoc (hunk 7 above)
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Usually the user would set the variable when running the `ansible-playbook` command.  (The user could specify it in the inventory file, but it probably is not helpful to mention that.)

Would it help to move the "see also" into the first sentence? "If you specify the `*openshift_checks_output_dir*` variable (see example usage of the `ansible-playbook` command below) [...]"

Or should I replace or supplement the "see also" at the end of the paragraph with something more explicit? "For example, use `ansible-playbook -i <inventory_file> ~/openshift-ansible/playbooks/openshift-checks/health.yml -e openshift_checks=sdn -e openshift_checks_output_dir=./ansible-output`.  See example usage of the `ansible-playbook` command below."

Comment (mburke5678) [reply admin_guide/diagnostics_tool.adoc]: Miciah. I didn't realize that you add the variable when you run the playbook. maybe:
_If you specify the `*openshift_checks_output_dir*` variable with the `ansible-playbook` command..._

File: admin_guide/diagnostics_tool.adoc (hunk 8 above)
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: Done.

File: admin_guide/diagnostics_tool.adoc (hunk 7 above)
Comment (Miciah) [reply admin_guide/diagnostics_tool.adoc]: All right, I've added "with the `ansible-playbook` command".  Thanks!

**Comment by kalexand-rh** [comment]
Thank you @mburke5678 and @Miciah! I'll merge when we open the 3.11 branch.

**Pull request merged**